tdrs-backend/tdpservice/reports/test/test_serializers.py

```diff
@@ -1,54 +1,89 @@
 """Test report serializers."""
-
+from django.core.exceptions import ValidationError
 import pytest
-from ..serializers import ReportFileSerializer
-from ..errors import ImmutabilityError
+
+from tdpservice.reports.errors import ImmutabilityError
+from tdpservice.reports.serializers import ReportFileSerializer
+from tdpservice.reports.validators import validate_file_extension
+
 
 @pytest.mark.django_db
-def test_serializer_with_valid_data(report):
+def test_serializer_with_valid_data(report_data):
     """If a serializer has valid data it will return a valid object."""
-    get_serializer = ReportFileSerializer(report)
-    create_serializer = ReportFileSerializer(data=get_serializer.data)
+    create_serializer = ReportFileSerializer(data=report_data)
+    create_serializer.is_valid(raise_exception=True)
     assert create_serializer.is_valid() is True
 
+
 @pytest.mark.django_db
-def test_serializer_increment_create(report):
+def test_serializer_increment_create(report_data, other_report_data):
     """Test serializer produces reports with correct version."""
-    get_serializer = ReportFileSerializer(report)
-    serializer_1 = ReportFileSerializer(data=get_serializer.data)
+    serializer_1 = ReportFileSerializer(data=report_data)
     assert serializer_1.is_valid() is True
     report_1 = serializer_1.save()
 
-    serializer_2 = ReportFileSerializer(data=get_serializer.data)
+    serializer_2 = ReportFileSerializer(data=other_report_data)
     assert serializer_2.is_valid() is True
     report_2 = serializer_2.save()
 
     assert report_2.version == report_1.version + 1
 
+
 @pytest.mark.django_db
 def test_immutability_of_report(report):
     """Test that report can only be created."""
-    try:
+    with pytest.raises(ImmutabilityError):
         serializer = ReportFileSerializer(
             report, data={
                 "original_filename": "BadGuy.js"
             },
-            partial=True)
+            partial=True
+        )
 
         serializer.is_valid()
         serializer.save()
-        raise Exception("Should not be able to update reports.")
-    except ImmutabilityError as err:
-        expected = "Cannot update, reports are immutable. Create a new one instead."
-        assert str(err) == expected
 
 
 @pytest.mark.django_db
-def test_created_at(report):
+def test_created_at(report_data):
     """If a serializer has valid data it will return a valid object."""
-    get_serializer = ReportFileSerializer(report)
-    create_serializer = ReportFileSerializer(data=get_serializer.data)
+    create_serializer = ReportFileSerializer(data=report_data)
     assert create_serializer.is_valid() is True
     report = create_serializer.save()
 
     assert report.created_at
+
+
+@pytest.mark.parametrize("file_name", [
+    'sample.txt',
+    'Sample',
+    'ADS.E2J.FTP4.TS06.txt',
+    'ADS.E2J.FTP4.TS06',
+    'ADS.E2J.FTP2.MS18',
+    'ADS.E2J.FTP1.TS278'
+])
+def test_accepts_valid_file_extensions(file_name):
```

Comment:
Using [parametrization](https://docs.pytest.org/en/6.2.x/parametrize.html) here allows us to run the same test against a number of inputs. Verbose output from test runs show a separate test being run for each input supplied:
```
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[sample.txt] PASSED                                                                                                          [ 16%]
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[Sample] PASSED                                                                                                              [ 33%]
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[ADS.E2J.FTP4.TS06.txt] PASSED                                                                                               [ 50%]
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[ADS.E2J.FTP4.TS06] PASSED                                                                                                   [ 66%]
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[ADS.E2J.FTP2.MS18] PASSED                                                                                                   [ 83%]
tdpservice/reports/test/test_serializers.py::test_accepts_valid_file_extensions[ADS.E2J.FTP1.TS278] PASSED                                                                                                  [100%]
```